# 异常

异常是程序运行中允许出现的，能被容忍的错误

## 处理异常


### 异常分类

`Throwable` 有两个子类

- `Error` 系统级别异常。通常是 `Java` 运行时内部错误或资源耗尽错误，这种错误通常都是致命的
- `Exception` 应用级别异常。分两大类
  - `RunTimeException` 运行时异常。通常出现异常的原因是代码的问题。比如，空指针，数组下标越界，类型转换错误
  - 其他异常。通常是程序代码没有问题，是运行时，环境复杂造成的异常。比如打开一个不存在的文件，进行网络通信时突然断网等

> “如果出现 **RuntimeException** 异常，那么就一定是你的问题“ 是一条相当有道理的规则

`Error` 和 `RunTimeException` 是非受查异常

其他异常是受查异常

编译器为受查异常提供**异常处理器（try-catch块）**


### 声明受检查的异常

如果方法可能抛出一些受查异常，应该在方法定义处声明这些异常

```java
public void functionName() throws AException, BException
```

不应该声明非受查异常（`Error` 的子类和 `RunnTimeException` 及其子类）


**子类重写父类方法时如何声明异常列表**

如果在子类中覆盖了超类的一个方法， 子类方法中声明的受查异常不能比超类方法中声明的异常更通用 （也就是说， 子类方法中可以抛出更特定的异常， 或者根本不抛出任何异常）

原因：父类引用的实际类型可能是子类对象，而父类引用调用方法只能抛出父类方法中声明的异常。不管对象的实际类型是什么都要保证异常能被捕捉，所以子类需要配合父类


### 抛出异常

```java
// throw + 异常对象
throw exception;
```


### 创建异常类

创建一个 `Exception` 或其子类的子类


## 捕捉异常


### 异常处理器

**如果一个异常没有被任何一个地方进行捕获，那程序就会终止执行**


异常处理器的构成

```java
try {
    ...
} catch(AException a) {
    ...
} catch(BException b) {
    ...
}
```

- 异常链中从上到下，`AException`<`Bexception` 或 `AException` 和 `Bexception` 毫无关系（异常的类型越来越通用或毫无关系）
- `catch` 列表中的异常变量隐含为 `final` 变量
- 抛出异常后，跳过 `try` 块中当前语句下的所有语句


### 再次抛出异常

在异常处理器中能再次抛出异常

这样做的目的通常是改变异常的类型


**异常包装技术**

```java
catch(SQLException e) {
    Throwable se = new ServletException ("database error");
	se.initCause(e);
	throw se;
}
```

改变异常的类型并再次抛出，当捕捉到异常时，能这样获取原始的异常

```java
Throwable e = se.getCause();
```


### finally 块

`finally` 块的格式（`catch` 块不是必须的）

```java
try {
    ...
} catch (Exception e){
    ...
} finally {
    ...
}
```

在执行 `try-catch-finally` 块中执行 `return` 前会执行 `finally` 里的语句

有关 `finally` 块的详细说明见 JVM，例如

- 在 `finally` 块中执行 `return`
- 在 `finally` 块中修改返回值，但无论在 finally 代码块中是否修改返回值，返回值都不会改变


## 带资源的 try 块


带资源的 `try` 语句（ `try-with-resources` ) 的最简形式

```java
// 列表中的资源用';'分割
try (Resource resource1 = ...; Resource resource2 = ...) {
	// work with resource
}
```

如果 `Resource` 实现了 `AutoCloseable `接口，在代码退出 `try` 块后会执行 `resource.close()` 方法（ `AutoCloseable ` 接口的方法）


> 如果 try 块抛出一个异常， 而且 close 方法也抛出一个异常，这就会带来一个难题。带资源的 try 语句可以很好地处理这种情况。原来的异常会重新抛出，而 close方法抛出的异常会“被抑制”，这些异常将自动捕获，并由 addSuppressed 方法增加到原来的异常。 如果对这些异常感兴趣， 可以调用 getSuppressed 方法，它会得到从 close 方法抛出并被抑制的异常列表


ps：带资源的 try 语句自身也可以有 catch 子句和一个 finally 子句。 这些子句会**在关闭资源之后**执行。

[finally块中使用return会抑制异常的冒泡传输和finally中抛异常应该如何处理](https://blog.csdn.net/qq_44861675/article/details/106353369)


## 获取堆栈轨迹

方式一

```java
Throwable t = new ThrowableO ;
StackTraceElement[] frames = t.getStackTrace();
for (StackTraceElement frame : frames)
	// analyze frame
```

方式二

```java
Map<Thread, StackTraceElement[]> map = Thread.getAl1StackTraces();
for (Thread t : map. keySet ()) {
    StackTraceElement[] frames = map.get(t);
    // analyze frames
}
```

`StackTraceElement` 类含有能够获得文件名和当前执行的代码行号的方法，还含有能够获得类名和方法名的方法


# 断言

> 测试程序时用 `System.out.println` 输出数据实现测试的策略执行时既费时又需要频繁去除（测试完成后需要手动去掉这些测试代码）
>
> 断言能选择性地启用测试以及打印测试信息


## 断言的概念和表达式

如果断言错误，抛出 `AssertError` 错误

如果断言条件后有表达式，表达式将作为错误信息传递到 `AssertError` 中

```java
assert 条件;

assert 条件 : 表达式;
```

`AssertError` 是非受查异常，断言出现的错误是不被期望的，不能被容忍的，一定需要被避免出现的错误

和 `Exception` 不同，断言错误在程序上线后是一定不能出现的


## 启用和禁用断言

默认断言被禁用。通过`JVM`参数控制断言的启用和禁用

```shell
# 启用所有断言
-ea
# 启用指定类或包下所有类的断言
-ea:OneClass -ea:com.example...

# 禁用断言
-da
# 禁用指定类或包下所有类的断言
-da:OneClass -ea:com.example...
```


## 断言实现参数检查

```java
int i = 3;

assert
```


## 使用断言的时机

- 断言用于开发和测试阶段
- 如果明确要求某个方法的参数一定符合某种规范，否则程序直接崩掉。这些情况下使用断言对参数进行检查


**断言和异常的使用场景**

- 异常用于处理已知的可能出现问题的地方，是项目上线后。例如，根据 `username` 从 `db` 获取数据，结果数据为空，应该抛类似 `UserNotFoundException`
- 断言用于处理预期中绝对不会出现的错误。例如某个方法的参数必须为正数，且正常情况下接收到的参数都是正数。


# 记录日志

> `rt.jar` 中 `java.util.logging` 是与日志相关的类

- `Logger` 日志记录器。真正打印日志的类
- `LoggerManager` 日志管理器。**日志记录器的工具类**，提供静态方法给日志记录器使用
- `LogRecord` 日志主体

可通过日志记录器/日志管理器设置启用/禁用日志，设置日志打印级别


## 日志记录器级别

7个级别，通常只关注前3个

`SEVERE`，`WARNING`，`INFO`

默认打印 `INFO` 及其以上级别的日志


## 获取日志记录器的方式

- 全局日志记录器
- 按类划分的日志记录器（每个类持有一个日志记录器）


## 记录日志的方法

- `logger.log` 通用方法
- `logger.logp` 比上面的方法更稳定。通常使用下面的方法间接调用这个方法，而不是直接调用
- `logger.server`、`logger.warning`、`logger.info` 对应日志级别的方法
- `logger.entering` 用于执行方法前，对类，方法，参数进行记录的日志方法
- `logger.exiting` 用于执行方法后，对类，方法，返回结果进行记录的日志方法
- `logger.throwing` 抛异常时使用的日志方法


## 日志管理器配置

`jdk` 的日志管理配置文件是 `jre/lib/1ogging.properties`

可通过系统变量指定其他的日志配置文件

`java -Djava.util.logging.config.file=configFile MainClass`

加载 `LoggerManager` 类时，执行静态代码块时会先获取自定义配置文件地址，如果没有就使用默认的配置文件

除此以外还能通过修改系统变量修改日志管理器或日志记录器


## 日志记录流程

这里说的是 `Logger` 和 `LoggerManager` 提供的日志服务中的默认日志记录流程

- 将日志本地化（根据本地信息对日志进行国际化处理）
- 日志过滤器过滤日志。将低于当前设置的日志级别的日志过滤掉不记录（ `boolean isLoggab1e(LogRecord record)` ）
- 将日志交给 `ConsoleHandler`
  - 日志格式化器格式化日志
  - 输出到 `System.err` 流中

本地化处理器，日志过滤器，日志处理器都可以通过日志记录器设置进去


# 调试技巧

- 使用 `IDE` 调试功能
- 使用 `JUnit` 单元测试
- 对类进行单元测试（在类的 `main` 方法中进行测试）
- 日志或主动打印堆栈轨迹信息并将信息持久化，分析这些信息


**拓展**

用 `-verbose` 标志启动 `Java` 虚拟机，有时候，这种方法有助于诊断由于类路径引发的问题。



